test/functorch/test_aotdispatch.py

```diff
@@ -2954,29 +2954,107 @@ def forward(self, arg0_1, arg1_1):
     _set_grad_enabled_1 = torch._C._set_grad_enabled(False)
     return (add_3,)""")
 
-    # TODO(tmanlaibaatar) properly support functionalizing HOO in
-    # predispatch tracing mode
-    @unittest.expectedFailure
+    def test_aot_export_predispatch_with_cond_nested(self):
+        class M(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+
+            def forward(self, x):
+                def true_fn(x):
+                    y = x.sin()
+                    y.add_(5)
+
+                    def true_true_fn(x):
+                        y = x.sin()
+                        y.add_(7)
+                        return y.sin()
+
+                    def true_false_fn(x):
+                        return x.cos()
+
+                    return torch.cond(y.cos().shape[0] > 5, true_true_fn, true_false_fn, [y.cos()])
+
+                def false_fn(x):
+                    z = x.cos()
+                    z.add_(6)
+                    return z.sin()
+
+                a = torch.cond(x.shape[0] > 4, true_fn, false_fn, [x])
+                return (a + 3, a + 4)
+
+        inp = torch.randn(2, 2)
+        gm, _ = aot_export_module(M(), [inp], trace_joint=False, pre_dispatch=True)
+        self.assertExpectedInline(str(gm.code).strip(), """\
+def forward(self, arg0_1):
+    _set_grad_enabled = torch._C._set_grad_enabled(False)
+    _set_grad_enabled_1 = torch._C._set_grad_enabled(False)
+    _set_grad_enabled_2 = torch._C._set_grad_enabled(False)
+    true_graph_0 = self.true_graph_0
+    false_graph_0 = self.false_graph_0
+    conditional = torch.ops.higher_order.cond(False, true_graph_0, false_graph_0, [arg0_1]);  true_graph_0 = false_graph_0 = arg0_1 = None
+    getitem = conditional[0];  conditional = None
+    add = torch.ops.aten.add.Tensor(getitem, 3)
+    add_1 = torch.ops.aten.add.Tensor(getitem, 4);  getitem = None
```

Comment:
I'm not convinced that the following case works. we should test it:
- cond where one of the branches calls some functional composite operation. We want to make sure that this composite operation doesn't get decomposed.

Reply:
Done